engine/access/rpc/backend/transaction_subscription_metadata.go

```diff
@@ -0,0 +1,254 @@
+//
+// No errors expected during normal operations.
+func newTransactionSubscriptionMetadata(
+	backendTransactions *backendTransactions,
+	txID flow.Identifier,
+	txReferenceBlockID flow.Identifier,
+	eventEncodingVersion entities.EventEncodingVersion,
+) *transactionSubscriptionMetadata {
+	return &transactionSubscriptionMetadata{
+		backendTransactions:  backendTransactions,
+		txResult:             &access.TransactionResult{TransactionID: txID},
+		eventEncodingVersion: eventEncodingVersion,
+		blocks:               backendTransactions.blocks,
+		collections:          backendTransactions.collections,
+		transactions:         backendTransactions.transactions,
+		txReferenceBlockID:   txReferenceBlockID,
+	}
+}
+
+// Refresh updates the transaction subscription metadata to reflect the latest state.
+//
+// Parameters:
+//   - ctx: Context for managing the operation lifecycle.
+//
+// Expected errors during normal operation:
+//   - [ErrBlockNotReady] if the block at the given height is not found.
+//   - codes.Internal if impossible to get transaction result due to event payload conversion failed
+//
+// All other errors are considered as state corruption (fatal) or internal errors in the refreshing transaction result
+// or when refreshing transaction status.
+func (tm *transactionSubscriptionMetadata) Refresh(ctx context.Context) error {
+	if err := tm.refreshCollection(); err != nil {
+		return err
+	}
+	if err := tm.refreshBlock(); err != nil {
+		return err
+	}
+	if err := tm.refreshTransactionResult(ctx); err != nil {
+		return err
+	}
+	if err := tm.refreshStatus(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// refreshTransactionReferenceBlockID sets the reference block ID for the transaction.
+//
+// If the reference block ID is unset, it attempts to retrieve it from storage.
+//
+// Parameters:
+//   - txReferenceBlockID: The reference block ID of the transaction.
+//
+// No errors expected during normal operations.
+func (tm *transactionSubscriptionMetadata) refreshTransactionReferenceBlockID() error {
+	// Get referenceBlockID if it is not set
+	if tm.txReferenceBlockID != flow.ZeroID {
+		return nil
+	}
+
+	tx, err := tm.transactions.ByID(tm.txResult.TransactionID)
+	if err != nil {
+		return fmt.Errorf("failed to lookup transaction by transaction ID: %w", err)
+	}
+	tm.txReferenceBlockID = tx.ReferenceBlockID
+	return nil
+}
+
+// refreshStatus updates the transaction's status based on its execution result.
+//
+// Parameters:
+//   - ctx: Context for managing the operation lifecycle.
+//
+// No errors expected during normal operations.
+func (tm *transactionSubscriptionMetadata) refreshStatus(ctx context.Context) error {
+	var err error
+
+	if tm.blockWithTx == nil {
+		if err = tm.refreshTransactionReferenceBlockID(); err != nil {
+			// transaction was not sent from this node, and it has not been indexed yet.
+			if errors.Is(err, storage.ErrNotFound) && tm.txReferenceBlockID == flow.ZeroID {
```

Comment:
looking at the logic, I don't think it's possible for both `err != nil` and `tm.txReferenceBlockID == flow.ZeroID` to be true at the same time
```suggestion
			if errors.Is(err, storage.ErrNotFound) {
```